File: PythonSRC/BlockChain.py
import os
from hashlib import sha256
from random import randint
from datetime import datetime
import json

class BlockChainGenerator():

    def __init__(self,fileName, newChain=True,difficultyLevel=2):
        self.fileName = fileName
        self.newChain = newChain
        self.difficultyLevel = difficultyLevel
        self.BlockChain = []
        self.HandleJson()
        self.AddBlock("GENESIS BLOCK") if self.BlockChain == [] else None

    def HandleJson(self):
        if not self.newChain:
            try:
                with open(self.fileName) as f:
                    self.BlockChain = json.load(f)
            except Exception as e:
                print(f"Error: {e}")
    
    #Create block
    def CreateBlock(self,data):
        block = {}
        block["data"] = data
        block["index"]= str(len(self.BlockChain))
        block["timeStamp"] = str(datetime.utcnow())
        block["previousHash"] = self.BlockChain[0]["currentHash"] if self.BlockChain != [] else "x"
        block["currentHash"], block["nounce"] = self.miner(
            block["data"]+block["index"]+block["timeStamp"]+block["previousHash"]
        )

        return block

    def miner(self,dataString):

        while True:
            nounce = str(randint(0,1E10))
            hash = sha256(str(dataString+nounce).encode()).hexdigest()
            if hash[:self.difficultyLevel]=="0"*self.difficultyLevel:
                return hash,nounce

    #add block to BlockChain
    def AddBlock(self,data):
        self.BlockChain = [self.CreateBlock(data)]+self.BlockChain 

        with open(self.fileName,"w") as f:
            json.dump(self.BlockChain,f)


#---------------------------------------------------
BlockChainFileName = "BlockChain.json"
Transaction = BlockChainGenerator(BlockChainFileName,newChain=False,difficultyLevel=3)
while True:
    Transaction.AddBlock(input("Enter Transaction ::: "))
    os.system(BlockChainFileName)

File: GoSRC/BlockChain.go
package main

import (
	"os"
	"os/exec"
	"crypto/sha256"
	"encoding/base64"
	"math/rand"
	"time"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
	"bufio"
)

//Using struct for Blockcahin generator

type BlockchainGenerator struct{

	FileName string
	NewChain bool
	DifficultyLevel	int
	BlockChain []Block
}

type Block struct{
	Data,Index, TimeStamp, PreviousHash,CurrentHash,Nounce string
}

func (b *BlockchainGenerator) HandleJson(){

	if !b.NewChain{
		filebytes,err := ioutil.ReadFile(b.FileName)
		if err != nil{
			fmt.Println("##",err) //Dont exit
		}else{

			var existingData []Block
			json.Unmarshal([]byte(filebytes),&existingData)
			b.BlockChain = existingData
		}
	}
}

//Create Block
func (b *BlockchainGenerator) CreateBlock(data string) Block{

	block := Block{}
	block.Data = data
	block.Index = strconv.Itoa(len(b.BlockChain))
	block.TimeStamp = time.Now().UTC().Format(time.RFC3339Nano)
	block.PreviousHash = func()string{
		if len(b.BlockChain) != 0 {
			return b.BlockChain[0].CurrentHash
		}else{
			return "x"
		}
	}()
	
	block.CurrentHash,block.Nounce = b.miner(block.Data+ block.Index+ block.TimeStamp+ block.PreviousHash)

	return block
}

func (b BlockchainGenerator) miner(dataString string) (string,string){

	for true{

		nounce := strconv.Itoa(rand.Intn(1e10))
		hash := sha256.Sum256([]byte(dataString+nounce))
		hash_str := base64.StdEncoding.EncodeToString(hash[:])

		if hash_str[:b.DifficultyLevel] == strings.Repeat("0",b.DifficultyLevel){

			return hash_str,nounce
		}
	}

	return "-","-"

}

// Add block

func (b *BlockchainGenerator) AddBlock(data string){

	b.BlockChain = append([]Block{b.CreateBlock(data)},b.BlockChain...)

	jsonFile,_ := json.MarshalIndent(b.BlockChain,""," ")
	_ = ioutil.WriteFile(b.FileName, jsonFile, 0644)
}


func init(){

	rand.Seed(time.Now().UnixNano()) //Random number linked to time
}


func main(){

	Transaction := BlockchainGenerator{

		FileName : "BlockChain.json",
		NewChain : false,
		DifficultyLevel: 2,

	}

	Transaction.HandleJson()

	if len(Transaction.BlockChain) == 0{

		Transaction.AddBlock("GENESIS BLOCK")
	}

	scanner := bufio.NewScanner(os.Stdin)

	for true{

		fmt.Printf("Enter Transaction ::: ")
		scanner.Scan()
		Transaction.AddBlock(scanner.Text())
		cmd := exec.Command("powershell","start","firefox",Transaction.FileName)
		cmd.Run()
	}


}